### PR TITLE
Remove `@wordpress/{components,plugins,date,dom,notices,keycodes,rich-text,redux-routine,block-serialization-default-parser,deprecated}` types due to bundled definitions in packages

### LINKED ISSUE #64541: [wordpress__components] /color-picker/index.d.ts does not support updated properties
### Discussed in https://github.com/DefinitelyTyped/DefinitelyTyped/discussions/62576

<div type='discussions-op-text'>

<sup>Originally posted by **mogmarsh** October  5, 2022</sup>
@types/wordpress__components/color-picker/index.d.ts references deprecated properties and does not support new properties.

See https://github.com/WordPress/gutenberg/commit/d2363e7ca7adf7f510c78a26ecbf24618b389b87 and https://github.com/WordPress/gutenberg/pull/35220/

From the changelog:
> The `ColorPicker` component property `onChangeComplete`, a function accepting a color object, was replaced with the property `onChange`, a function accepting a  string on  ([#35220](https://github.com/WordPress/gutenberg/pull/35220)).
> The property `disableAlpha`, was removed from the `ColorPicker` component. Use the new opposite property `enableAlpha` instead ([#35220](https://github.com/WordPress/gutenberg/pull/35220)).

</div>

### PR DESCRIPTION
**Pull request by rafaucau**
This PR removes types that are already bundled in the packages.

- [@wordpress/components](https://www.npmjs.com/package/@wordpress/components?activeTab=code)
- [@wordpress/plugins](https://www.npmjs.com/package/@wordpress/plugins?activeTab=code)
- [@wordpress/date](https://www.npmjs.com/package/@wordpress/date?activeTab=code)
- [@wordpress/dom](https://www.npmjs.com/package/@wordpress/dom?activeTab=code)
- [@wordpress/notices](https://www.npmjs.com/package/@wordpress/notices?activeTab=code)
- [@wordpress/keycodes](https://www.npmjs.com/package/@wordpress/keycodes?activeTab=code)
- [@wordpress/rich-text](https://www.npmjs.com/package/@wordpress/rich-text?activeTab=code)
- [@wordpress/redux-routine](https://www.npmjs.com/package/@wordpress/redux-routine?activeTab=code)
- [@wordpress/block-serialization-default-parser](https://www.npmjs.com/package/@wordpress/block-serialization-default-parser?activeTab=code)
- [@wordpress/deprecated](https://www.npmjs.com/package/@wordpress/deprecated?activeTab=code)

---

Resolves: #64541

### CODE REVIEW

**Comment by DangerBotOSS** [comment]
## Formatting

The following files are not formatted:
1. types/wordpress__editor/components/autocompleters.d.ts
1. types/wordpress__block-editor/components/media-placeholder.d.ts
1. types/wordpress__block-editor/components/block-controls.d.ts
1. types/wordpress__block-editor/components/alignment-toolbar.d.ts
1. types/wordpress__block-editor/components/rich-text.d.ts

Consider running `pnpm dprint fmt` on these files to make review easier.
<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 72c5f1ee43ad7bc5969897595a071db4b4a9809b
</p>

**Review by sarayourfriend — APPROVED** [review]
LGTM. I've gone through each of the listed packages and confirmed they all indeed export their own types now.

File: types/wordpress__block-editor/components/block-format-controls.d.ts
Comment (sarayourfriend) [inline types/wordpress__block-editor/components/block-format-controls.d.ts:1]: Just one nit, I don't know if accessing the types this way is safe. Specifically, whether Gutenberg's public API rules would indicate that this is a _stable_ way to reference these types. It might create a hard-pin on a specific version that wouldn't be obvious (usually moving files around isn't cause for even a patch version bump in semver).

Comment (rafaucau) [reply types/wordpress__block-editor/components/block-format-controls.d.ts:1]: This is the only option I was aware of. The types are not exported from the main package file.
Maybe @gziolo knows the answer?

Comment (mikeybinns) [reply types/wordpress__block-editor/components/block-format-controls.d.ts:1]: I did it differently on my PR, I used Parameters to avoid this issue, the only type I had to import like this was Popover. Perhaps they should be imported that way instead? Wordpress doesn't explicitly export its types for Props, only for the full components.
https://github.com/DefinitelyTyped/DefinitelyTyped/pull/69143

Comment (gziolo) [reply types/wordpress__block-editor/components/block-format-controls.d.ts:1]: The list of props for React components definitely sounds like public API. I have no insights why would they be not exposed from the root of the package.

Comment (rafaucau) [reply types/wordpress__block-editor/components/block-format-controls.d.ts:1]: Unfortunately, in your approach, for example, the `typeof slot` is of type `unknown`. I tried with `React.ComponentProps<typeof Slot>`, but also the result is `unkwown`.
image

Direct import of types:
image

Comment (mikeybinns) [reply types/wordpress__block-editor/components/block-format-controls.d.ts:1]: Ah I didn't notice that, well spotted

Comment (mikeybinns) [reply types/wordpress__block-editor/components/block-format-controls.d.ts:1]: Issue submitted: https://github.com/WordPress/gutenberg/issues/60193

**Comment by gziolo** [comment]
Overall, great effort to use all the exposed types from the updated WordPress packages. There might be still a few smaller packages that could natively offer types without too much hassle.

File: types/wordpress__block-editor/components/block-format-controls.d.ts
Comment (rafaucau) [reply types/wordpress__block-editor/components/block-format-controls.d.ts:1]: So the plan is:

1. Merge this PR as it is
2. Create PR to @wordpress/components that will export types from root
3. Wait for merge and release
4. Create second PR to DT that will update imports

Do you agree @sarayourfriend?

Comment (gziolo) [reply types/wordpress__block-editor/components/block-format-controls.d.ts:1]: Sounds great!

Comment (sarayourfriend) [reply types/wordpress__block-editor/components/block-format-controls.d.ts:1]: Yes, sounds good to me.

Comment (rafaucau) [reply types/wordpress__block-editor/components/block-format-controls.d.ts:1]: Ok. Can you merge this PT then?

Comment (sarayourfriend) [reply types/wordpress__block-editor/components/block-format-controls.d.ts:1]: Merging this is blocked on:

> This PR touches some part of DefinitelyTyped infrastructure, so a DT maintainer will need to review it. This is rare — did you mean to do this?

And from the checklist:

> A DT maintainer needs to approve changes which affect DT infrastructure ([notNeededPackages.json](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/69133/files/af6acf82514e343c8d5d16e4752317a97e12c892#diff-a275c2eae7b8f788a52327e76809026ee1bc5ea614393806c8f3cbab07202621))

It's been a while since I've done this, so I'd check out the contributing documentation for the repo to see if there are instructions on getting that review from a DT maintainer :+1:

Comment (rafaucau) [reply types/wordpress__block-editor/components/block-format-controls.d.ts:1]: The documentation does not say exactly how to do this. There is only [information](https://github.com/DefinitelyTyped/dt-mergebot/blob/master/docs/dt-mergebot-lifecycle.svg) that the maintainer must accept it.
I see from other pull requests that some of the maintainers are @peterblazejewicz, @sheetalkamat.

Comment (sarayourfriend) [reply types/wordpress__block-editor/components/block-format-controls.d.ts:1]: Okay, well that graph makes it look like with my approval this PR will be in the DT maintainer review queue, so just hold tight while they get to it, I guess.

**Comment by rafaucau** [comment]
It looks like the problem with component type detection (https://github.com/DefinitelyTyped/DefinitelyTyped/pull/69133#discussion_r1538999762) is coming from my IDE.
PHPStorm:
image

VSCode:
image

Last time I had looked, the problem was also in the VSC, but today it is ok. In that case, I'll be continuing this PR using VSC.

**Comment by mikeybinns** [comment]
Okay, great, so just to confirm, the plan is now to update your PR to remove the deep imports and instead use React.ComponentProps to get the props for each component?

I saw you added me to your repo, were you looking for help with this @rafaucau? Or was that just for assistance with the bug you mentioned above?

**Comment by rafaucau** [comment]
> @mikeybinns  I saw you added me to your repo, were you looking for help with this @rafaucau? Or was that just for assistance with the bug you mentioned above?

Initially with bug mentioned above. But this has already been solved.

But there is one last challenge. Somehow the `WPCompleter` type has to be extracted so that it is possible to pass the generic like `EditorInserterItem` or `User`.

https://github.com/DefinitelyTyped/DefinitelyTyped/blob/145a12bbdd8c59a88428902add4791fc76e75d6a/types/wordpress__editor/components/autocompleters.d.ts#L1-L6

This is how I managed to extract the type of `WPCompleter`, but unfortunately then it is not possible to pass the generic:
```ts
type Completer = React.ComponentProps<typeof Autocomplete>['completers'][number];
```
So close...
image

I have no idea at this moment. Can you help?

**Comment by mikeybinns** [comment]
> But there is one last challenge. Somehow the `WPCompleter` type has to be extracted so that it is possible to pass the generic like `EditorInserterItem` or `User`.

I don't think it's possible to both extract a type from an array and modify it with a generic, so for that one, you might need to just use that deep import 🤷 This is made clearer if you imagine the type was given a generic in its implemenation.

e.g.
```ts
type ThingWithGeneric<T> = {test: T};
type ArrayType = ThingWithGeneric<string>[]
type Test = ArrayType[number] // ThingWithGeneric<string>, string has already been passed to it.
```

**Comment by sarayourfriend** [comment]
I think it's a fine stop-gap to accept it will be `WPCompleter<any>` for now, and submit a PR to Gutenberg that turns the autocompleter component function into a generic so that this becomes possible.

```ts
import React from 'react'

type WPCompleter<T = any> = {
    name: T
}

type Props<T> = {
    completers: WPCompleter<T>[]
}

const Component = <T, >(props: Props<T>) => {
    return <div />
}

type Completer<T> = React.ComponentProps<typeof Component<T>>['completers'][number]
```

[Here's the TS playground link to try it out](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAJQKYEMDG8BmUIjgcilQ3wChSYBPMJOAdQAUBhXMAGyRiSgB4AVOAF44KAHaUAfELgBvUnAVxRKEEgBccPqQC+5KjTgMcYAM78pwuYrhpWHLlBMbGLcPe7mA2gF0d5W6Im8K6Qokii8ML8ADRwEgAUYMZOhsnmAJRCUlaKRDAArlCicDwAJsAAbnAA9BJ+FNS0Ie68fBaIxDAAdCEQYRFGEKY8+kgQmHC9-TDmEp74tm6c3Cb43p6i+SAARty+DQYAqibczctQ0mcOPEFQwKIA5nXV1YoAegD8B7QAoiAQACtgFduJc7OceJsdtxnq8FJ9-H0gnAuEFnMxwddbvcntJ4jIlCp1AQUCZSph8HBtCITHBjqdMdx0qQXtYEUA)

I haven't worked with React in almost three years, so I don't know whether React with TypeScript works well with these generic prop types (I seem to recall them being finicky).

**Review by sarayourfriend — APPROVED** [review]

**Comment by mikeybinns** [comment]
> I haven't worked with React in almost three years, so I don't know whether React with TypeScript works well with these generic prop types (I seem to recall them being finicky).

I think they work okay, I haven't had issues on my end, it's just a bit of uncommon syntax. I have a separate PR trying to add something similar for SelectControl, so I could make a PR for this too.

They may also make an exception for exporting the completer type, like they have done with IconType? I think best option may be submit an issue first explaining the issue and get their recommendation for their preferred fix :)

File: types/wordpress__block-editor/components/url-popover.d.ts
Comment (mikeybinns) [inline types/wordpress__block-editor/components/url-popover.d.ts:5]: Is there a reason this was changed from interface to type? Some users might be extending the interface and this would break that.

File: types/wordpress__blocks/api/categories.d.ts
Comment (mikeybinns) [inline types/wordpress__blocks/api/categories.d.ts:8]: IconType includes JSX.Element, so this can be simplified

**Comment by peterblazejewicz** [comment]
2c: I'd suggest splitting this one into two (removals stays, amendments goes into new PR via branch rename). It looks like removals are non controversial. Just an opinion

File: types/wordpress__block-editor/components/url-popover.d.ts
Comment (rafaucau) [reply types/wordpress__block-editor/components/url-popover.d.ts:5]: Yes. There is `An interface can only extend an object type or intersection of object types with statically known members.ts(2312)` error with `interface`

Comment (mikeybinns) [reply types/wordpress__block-editor/components/url-popover.d.ts:5]: Okay ignore this then :)

**Review by mikeybinns — APPROVED** [review]

**Review by RyanCavanaugh — APPROVED** [review]

**Comment by rafaucau** [comment]
Ready to merge

**Comment by margolisj** [comment]
Thanks @rafaucau  and @mikeybinns. I appreciate all the typing!

**Pull request merged**
